File: solutions/erlang/custom-set/1/src/custom_set.erl
-module(custom_set).

-export([add/2, contains/2, difference/2, disjoint/2, empty/1, equal/2, from_list/1,
         intersection/2, subset/2, union/2]).

-type elem() :: integer().
-type set() :: integer().

-spec add(Elem :: elem(), Set :: set()) -> set().
add(Elem, Set) ->
    union(Set, 1 bsl Elem).

-spec contains(Elem :: elem(), Set :: set()) -> boolean().
contains(Elem, Set) ->
    not empty(intersection(Set, 1 bsl Elem)).

-spec difference(Set1 :: set(), Set2 :: set()) -> set().
difference(Set1, Set2) ->
    Set1 bxor intersection(Set1, Set2).

-spec disjoint(Set1 :: set(), Set2 :: set()) -> boolean().
disjoint(Set1, Set2) ->
    empty(intersection(Set1, Set2)).

-spec empty(Set :: set()) -> boolean().
empty(Set) ->
    equal(Set, 0).

-spec equal(Set1 :: set(), Set2 :: set()) -> boolean().
equal(Set1, Set2) ->
    Set1 == Set2.

-spec from_list(List :: [elem()], Set :: set()) -> set().
from_list([], Set) ->
    Set;
from_list([X | Xs], Set) ->
    from_list(Xs, add(X, Set)).

-spec from_list(List :: [elem()]) -> set().
from_list(List) ->
    from_list(List, 0).

-spec intersection(Set1 :: set(), Set2 :: set()) -> set().
intersection(Set1, Set2) ->
    Set1 band Set2.

-spec subset(Set1 :: set(), Set2 :: set()) -> boolean().
subset(Set1, Set2) ->
    equal(intersection(Set1, Set2), Set1).

-spec union(Set1 :: set(), Set2 :: set()) -> set().
union(Set1, Set2) ->
    Set1 bor Set2.

File: solutions/go/custom-set/1/custom_set.go
package stringset

import (
	"fmt"
	"strings"
)

type Set map[string]struct{}

func New() Set {
	return Set{}
}

func NewFromSlice(l []string) Set {
	set := New()
	for _, elem := range l {
		set.Add(elem)
	}
	return set
}

func (s Set) String() string {
	elems := make([]string, 0, len(s))

	for elem := range s {
		elems = append(elems, fmt.Sprintf("\"%s\"", elem))
	}

	return fmt.Sprintf("{%s}", strings.Join(elems, ", "))
}

func (s Set) IsEmpty() bool {
	return len(s) == 0
}

func (s Set) Has(elem string) bool {
	_, ok := s[elem]

	return ok
}

func (s Set) Add(elem string) {
	s[elem] = struct{}{}
}

func Subset(s1, s2 Set) bool {
	for elem := range s1 {
		if !s2.Has(elem) {
			return false
		}
	}

	return true
}

func Disjoint(s1, s2 Set) bool {
	for elem := range s1 {
		if s2.Has(elem) {
			return false
		}
	}

	return true
}

func Equal(s1, s2 Set) bool {
	if len(s1) != len(s2) {
		return false
	}

	for elem := range s1 {
		if !s2.Has(elem) {
			return false
		}
	}

	return true
}

func Intersection(s1, s2 Set) Set {
	set := New()

	for elem := range s1 {
		if s2.Has(elem) {
			set.Add(elem)
		}
	}

	return set
}

func Difference(s1, s2 Set) Set {
	set := New()

	for elem := range s1 {
		if !s2.Has(elem) {
			set.Add(elem)
		}
	}

	return set
}

func Union(s1, s2 Set) Set {
	set := New()

	for elem := range s1 {
		set.Add(elem)
	}

	for elem := range s2 {
		set.Add(elem)
	}

	return set
}
